File: leetcode/array/best_time_to_buy_and_sell_stock.cc
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

// time complexity: O(n)

#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int min_p = INT32_MAX, profit = 0;
        for (const auto &p : prices) {
            min_p = min(min_p, p);
            profit = max(profit, p - min_p);
        }
        return profit;
    }
};

File: leetcode/array/best_time_to_buy_and_sell_stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# time complexity: O(n)

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0
        min_price = float('inf')
        for p in prices:
            min_price = min(min_price, p)
            profit = max(profit, p - min_price)
        return profit